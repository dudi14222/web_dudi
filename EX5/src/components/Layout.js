import React from 'react';
import { 
    Header,
    Body
} from '../components'
import { connect } from 'react-redux';
import classNames from 'classnames';
import authService from '../services/authenticationService';
import T from 'i18n-react';
import heb from '../translations/heb.json';
import en from '../translations/en.json';

    
class Layout extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            menuState: false,            
            isEnLanguage: localStorage.lang !== 'he'
        }
        T.setTexts(localStorage.lang === 'he' ? heb:en);
    }
 
  open = () =>  {
    this.setState({
      menuState: true
    })
  }

  switchLanguage = () =>{
      if(this.state.isEnLanguage){ 
          T.setTexts(heb);         
          this.setState({ isEnLanguage: false});
          localStorage.setItem("lang", "he");
      }
      else{    
          T.setTexts(en);      
          this.setState({ isEnLanguage: true});
          localStorage.setItem("lang", "en");
      }
  }

  close = (e) => {
    if (this.state.menuState) {
      this.setState({
        menuState: false
      })
    } 
  }

  render() { 
    const {         
        open, 
        close, 
        switchLanguage,
        state: { menuState }, 
        props: { 
            className,
            children,
            subHeader,
            showHeaderButton          
        }
    } = this;
    const headerButton = authService.isLogedIn ?         
        {   
            name: {key: 'Cart', nItems: this.props.items.length},
            path: '/cart'
        }
        :
        {   
            name: {key: 'LogIn'},
            path: '/login'
        };
    return (
        <div className={classNames(["App", className ])} onClick={() => close()}>    
            <Header showHeaderButton={showHeaderButton} headerButton={headerButton} menuConfig={{ menuState, open, switchLanguage }} >
                {subHeader}
            </Header>
            <Body>
                { children }
            </Body>              
        </div>
    )
  }
}

const mapStateToProps = ({cartReducer: {items}}) => ({
    items: items
})

export default connect(mapStateToProps, null)(Layout);

