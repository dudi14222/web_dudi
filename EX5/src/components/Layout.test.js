
import {Layout} from './Layout';
import React from 'react';
import { MemoryRouter } from 'react-router-dom'
import { Header, Body, Menu } from './';
import { mount, shallow } from "enzyme";
import renderer from 'react-test-renderer';

global.window = {}
import localStorage from 'mock-local-storage'
window.localStorage = global.localStorage

describe('Layout outer tests', () => {

    let props;
    let mountedlayoutComponent;

    const layoutComponent = () => {
        if (!mountedlayoutComponent) {
            mountedlayoutComponent = mount(
                <MemoryRouter>
                    <Layout  />
                </MemoryRouter>
            );
        }
        return mountedlayoutComponent
    }

    beforeEach(() => {
        props = {};
        mountedlayoutComponent = undefined;
    });
    
    it('is Layout', () => {
        const LayoutComponent = <Layout />;
        expect(LayoutComponent.type).toBe(Layout)
    })

    it("always renders a div", () => {
        const div = layoutComponent().find("div");
        expect(div.length).toBeGreaterThan(0);
    });

    it("has Header", () => {
        expect(layoutComponent().find(Header).length).toBe(1)
    })
    
    it("has Body Component", () => {
        expect(layoutComponent().find(Body).length).toBeGreaterThan(0)
    })
    
    it('Layout triggers menu open when clicked on menu', () => {
        const hamburgerButton = layoutComponent().find(Menu).find('.hamburger-icon');
        hamburgerButton.simulate('click');        
        expect(layoutComponent().find(Menu).props().menuConfig.menuState).toBe(true);
        hamburgerButton.simulate('click');
        expect(layoutComponent().find(Menu).props().menuConfig.menuState).toBe(false);
    })

    
})

describe('Layout snapshot tests', () => {
    it('test snapshot', () => {
        const tree = renderer.create(
            <MemoryRouter>
                <Layout />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    })
})

describe('Layout shallow tests', () => {

    it(`Layout has menu state and it's false`, () => {
        const layout = shallow(<Layout />)
        expect(layout.state().menuState).toBe(false);
    })

    it(`Layout manages menu close on click`, () => {
        const layout = shallow(<Layout />)
        layout.state().menuState = true;
        expect(
            layout.simulate('click')
                .state()
                    .menuState
        ).toBe(false);
    })

})