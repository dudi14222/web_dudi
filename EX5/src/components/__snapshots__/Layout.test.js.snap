// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout snapshot tests test snapshot 1`] = `
<div
  className="App"
  onClick={[Function]}
>
  <header
    className="App-header"
  >
    
    <div
      className="App-menubar"
    >
      <label
        className="hamburger-icon fa fa-bars"
        onClick={[Function]}
      />
      <div
        className="col-md-2 sidenav text-center closed"
      >
        <div
          className="parent-side-SideBar"
        >
          <div
            className="row"
          >
            <div
              className="logo-brand"
            >
              <img
                className="logo"
                src="/images/logo.jpg"
                style={
                  Object {
                    "height": "100px",
                    "width": "100px",
                  }
                }
              />
            </div>
          </div>
          <ul
            className="nav nav-pills nav-stacked"
          >
            <li>
              <a
                className="page-link"
                href="/"
                onClick={[Function]}
              >
                <span>
                  Home
                </span>
              </a>
            </li>
            <li>
              <a
                className="page-link"
                href="/about"
                onClick={[Function]}
              >
                <span>
                  About
                </span>
              </a>
            </li>
            <li>
              <a
                className="page-link"
                href="/products"
                onClick={[Function]}
              >
                <span>
                  Products
                </span>
              </a>
            </li>
            <li>
              <a
                className="page-link"
                href="/contact"
                onClick={[Function]}
              >
                <span>
                  Contact
                </span>
              </a>
            </li>
            <li>
              <a
                className="page-link"
                href="/login"
                onClick={[Function]}
              >
                <span>
                  Log In
                </span>
              </a>
            </li>
            <li>
              <button
                className="btn-change btn btn-primary"
                onClick={[Function]}
                type="button"
              >
                Hebrew
              </button>
            </li>
          </ul>
          <div
            className="social-media"
          >
            <ul
              className="list-inline"
            >
              <li>
                <a
                  className="icoFacebook"
                  href="https://www.facebook.com/dudi.hamdi"
                  target="_blank"
                  title="Facebook"
                >
                  <i
                    className="fa fa-facebook"
                  />
                </a>
              </li>
              <li>
                <a
                  className="icoTwitter"
                  href="https://twitter.com/?lang=en"
                  target="_blank"
                  title="Twitter"
                >
                  <i
                    className="fa fa-twitter"
                  />
                </a>
              </li>
              <li>
                <a
                  className="icoGoogle"
                  href="https://plus.google.com/discover?hl=iw"
                  target="_blank"
                  title="Google +"
                >
                  <i
                    className="fa fa-google-plus"
                  />
                </a>
              </li>
              <li>
                <a
                  className="icoLinkedin"
                  href="https://www.linkedin.com/"
                  target="_blank"
                  title="Linkedin"
                >
                  <i
                    className="fa fa-linkedin"
                  />
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      className="App-subheader"
    />
  </header>
  <div
    className="col-md-12 app-body"
  />
</div>
`;
